elements/pl-order-blocks/dag_checker.py

```diff
@@ -1,72 +1,50 @@
-
-class AnswerState:
-    def __init__(self):
-        self.valid = set()  # a block is valid if all blocks that it depends on have appeared before it
-        self.current_group = None
-        self.current_group_len = 0
-
-    def begin_group(self, group_id):
-        self.current_group = group_id
-
-    def end_group(self):
-        self.current_group = None
-        self.current_group_len = 0
-
-    def set_valid(self, stmt_id):
-        self.valid.add(stmt_id)
-        if self.current_group is not None:
-            self.current_group_len += 1
-
-    def is_valid(self, stmt_id):
-        return stmt_id in self.valid
-
-    def get_current_group_blocks(self):
-        return self.current_group_len
+from collections import Counter
+import networkx as nx
+
+
+def check_topological_sorting(order, graph):
+    seen = set()
+    for i, node in enumerate(order):
+        if node is None or not all([u in seen for (u, _) in graph.in_edges(node)]):
+            return i
+        seen.add(node)
+    return len(order)
+
+
+def check_grouping(order, group_belonging):
+    group_sizes = Counter(group_belonging.values())
+    cur_group = None
+    cur_group_size = None
+    for i, node in enumerate(order):
+        group_id = group_belonging.get(node)
+        if group_id is not None and cur_group is None:
+            cur_group = group_id
+            cur_group_size = 1
+        elif group_id is None and cur_group is not None:
+            return i
+        elif group_id is not None and cur_group is not None:
+            if group_id == cur_group:
+                cur_group_size += 1
+                if cur_group_size == group_sizes[cur_group]:
+                    cur_group = None
+                    cur_group_size = None
+            else:
+                return i
```

Comment:
an alternative way to write this for loop would be: 

```
    for i, node in enumerate(order):
        group_id = group_belonging.get(node)
        if group_id == cur_group:
            if group_id is None:
                continue
            cur_group_size += 1
            if cur_group_size == group_sizes[cur_group]:
                cur_group = None
                cur_group_size = None
        else:
            if group_id is not None and cur_group is None:
                cur_group = group_id
                cur_group_size = 1
            else:
                return i
```

I don't really love either way, I'm definitely open to suggestions on how to make this prettier.

Reply:
No one has proposed a simpler way to write this code, so I'm going to resolve this. 

Maybe I'll think something up later.